clang/test/CIR/CodeGen/asm.c

```diff
@@ -1,12 +1,32 @@
 // RUN: %clang_cc1 -triple x86_64-unknown-linux-gnu -fclangir-enable -emit-cir %s -o %t.cir
 // RUN: FileCheck --input-file=%t.cir %s
 
-//CHECK: cir.asm(x86_att, {""})
+//CHECK: cir.asm(x86_att, {"" ""})
 void empty1() {
   __asm__ volatile("" : : : );
 }
 
-//CHECK: cir.asm(x86_att, {"xyz"})
+//CHECK: cir.asm(x86_att, {"xyz" ""})
 void empty2() {
   __asm__ volatile("xyz" : : : );
+}
+
+//CHECK: cir.asm(x86_att, {"" "=*m,m"})
+void t1(int x) {
+  __asm__ volatile("" : "+m"(x));
+}
+
+//CHECK: cir.asm(x86_att, {"" "m"})
+void t2(int x) {
+  __asm__ volatile("" : : "m"(x));
+}
+
+//CHECK: cir.asm(x86_att, {"" "=*m"})
+void t3(int x) {
+  __asm__ volatile("" : "=m"(x));
```

Comment:
My current concern here is that we're not "officially" catching the input operands as part of the `cir.asm` operation, is this because the operand support is coming in another PR? If so, totally fine. Regarding the syntax, given:

```
__asm__ volatile("" : "=&r"(x), "+&r"(x));
```

My idea was to bound them like:

```
cir.asm(x86_att, {""}, %x : ["=&r", "+&r"], ...)
```

(So for this round (assuming operands come next) you could omit the operand)
In this way, whatever integer you use in the asm string, could map to the operand index starting at position 2. But I don't know how that complicates later LLVM lowering, thoughts?

Reply:
Yes, no operads so far, they'll come in future

> In this way, whatever integer you use in the asm string, could map to the operand index starting at position 2. But I don't know how that complicates later LLVM lowering, thoughts?

Sorry, didn't get you :( Could you please elaborate a little?

Reply:
Sure! Taking this from the tests:

```
cir.asm(x86_att, {"" "=&r,=&r,1"})
```

Here, my reading is that `1` above specifies that the constrains are referring to "the first operand" (though maybe it should have been `0`?). When you look into my suggestion:

```
cir.asm(x86_att, {""}, %x : ["=&r", "+&r"], ...)
```

What I'm really saying here is: instead of encoding `1`, we can look at the operand index of `%x` in the `cir.asm` operation to decide what to do when encoding the LLVM representation.

Does that make sense?

Reply:
Here the reading is that the **third** `output` operand has so-called early-clobber constraint.  Well, we could add a flag for this purpose, but I would not do it, if it's possible. 
1) It will make our code more different from the traditional `CodeGen`. 
2) there is one more place where the index can be added  in the constraint, so we will need to add one more flag besides early-clobber
3) I don't think we'll ever analyze inline asm in CIR - so the less it's differ from the original one the better

Reply:
@bcardosolopes  let's continue here as well)

Reply:
Thanks for your patience! 

> I don't think we'll ever analyze inline asm in CIR - so the less it's differ from the original one the better

This is a good point. Since my biggest concern here is to provide some degree of readability, given the original grammar for this:

> asm asm-qualifiers ( AssemblerTemplate
>                       : OutputOperands
>                       : InputOperands
>                       : Clobbers
>                       : GotoLabels)

I'd prefer if we are still closer to C/C++ than LLVM, so I'd expect something closer to:

```
cir.asm(x86_att, {""}, out {"=&r"(%x), ...}, in {"+&r"(%x), ...}, clobbers {...}, goto_labels {...})
or
cir.asm(x86_att, {""}, out ["=&r"(%x), ...], in ["+&r"(%x), ...], clobbers [...], goto_labels [...])
or
???
````

This is not a full fledged syntax, but you get the idea. Then as part of LLVM lowering we could do more of the lower level transformations: since we are probably not interested in analyzing inline asm in CIR, is there any need to stick the `1` this early in the game? Perhaps I'm missing something.

Reply:
@bcardosolopes  I was thinking about your words and tried to find a solution, but I failed. 

Well, the main point is - all that you see in the CIR output - even now - is already preprocessed strings, i.e. we take operands, analyze them, simplify and change constraint strings, insert something in the specific positions and etc. More than, there are functions that depend on `clang::Expr` and `clang::Stmt`, and these functions also change constraint strings. 

You could take a look at the traditional [codegen](https://github.com/llvm/llvm-project/blob/main/clang/lib/CodeGen/CGStmt.cpp#L2784). Just several lines, as an example, what the logic can be hidden behind. Yes, it's dirty :) But the problem is - the constraint string is already preprocessed, meaning there is an order which we also need to follow.

This is actually a job that have to be done exactly here, unless we can have an access to clang expressions and statements in the lowering stage. I understand (and admire) your desire to make things closer to C/C++ language, but I think it's either not doable or will cost too much. 

Again, we could find a way and don't inject all this annoying `1` . But in this case we would need to have a list of attributes, or flags or optional indexes for each constraint. I.e. we will have to reflect all the checks that are done in the code now in some array of flags and add it in the operation. And we would have to print it to make the operation parseable later, so  the readability would be decreased significantly.

As an idea, we could add a string that you want to see - just for pretty printing and nothing else.

Reply:
> This is actually a job that have to be done exactly here, unless we can have an access to clang expressions and statements in the lowering stage. I understand (and admire) your desire to make things closer to C/C++ language, but I think it's either not doable or will cost too much.

Thanks for more detailed background info. We could extend and pass AST nodes around in order to get that at the lowering stage (like we do for other ops), but I agree it might be too much given we have nothing else to gain from it.

Let's move on with this PR. My remaining request here is that you expand the docs to reflect this example:
```
//CHECK: cir.asm(x86_att, {"" "=&r,=&r,1"})
void t4(int x) {
  __asm__ volatile("" : "=&r"(x), "+&r"(x));
```

and as you move on with more PRs, make sure to enhance the docs. No need to explain again whatever the LLVM docs already do, but it's important to me that users need not to go look into codegen to graps the big picture.